src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2626,210 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    if (hasCheckedOverridden(mt))
+    {
+        return hasOverridden(mt);
+    }
+
+    BOOL hasOverriddenMethod = FALSE;
+
+    // Do we have the overridden method in current Type?
+    if (HasOverriddenMethodWorker(mt, classMT, methodSlot))
+    {
+        hasOverriddenMethod = TRUE;
+    }
+    else
+    {
+        // Now we need to do a depth-first-search on the instance fields.
+        Stack<ApproxFieldDescIterator> fieldIterStack;
+        fieldIterStack.push(ApproxFieldDescIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS));
+        while (!fieldIterStack.empty())
+        {
+            ApproxFieldDescIterator& iter = fieldIterStack.top();
+            FieldDesc* pField = iter.Next();
+
+            while (pField != NULL
+                && pField->GetFieldType() != ELEMENT_TYPE_VALUETYPE
+                && pField->GetFieldType() != ELEMENT_TYPE_CLASS)
+            {
+                // We cannot find overridden method in non-class types.
+                pField = iter.Next();
+            }
+
+            if (pField == NULL)
+            {
+                fieldIterStack.pop();
+                continue;
+            }
+
+            // Try to find the specified overridden method in current field type.
+            MethodTable* fieldMethodTable = pField->LookupFieldTypeHandle().GetMethodTable();
+            if (hasCheckedOverridden(fieldMethodTable))
+            {
+                // If we have checked this field type previously, it implies this field type
+                // does not have the specified overridden method.
+                continue;
+            }
+
+            if (HasOverriddenMethodWorker(fieldMethodTable, classMT, methodSlot))
+            {
+                setHasOverridden(fieldMethodTable);
+                setHasCheckedOverridden(fieldMethodTable);
+
+                hasOverriddenMethod = TRUE;
+
+                // If the specified overridden method found, terminate the DFS immediately.
+                break;
+            }
+
+            setHasCheckedOverridden(fieldMethodTable);
+
+            // Go for subfield type.
+            fieldIterStack.push(ApproxFieldDescIterator(fieldMethodTable, ApproxFieldDescIterator::INSTANCE_FIELDS));
+        }
+    }
+
+    // Cache the result.
+    setHasCheckedOverridden(mt);
+    if (hasOverriddenMethod)
+    {
+        setHasOverridden(mt);
+    }
+
+    return hasOverriddenMethod;
+}
+
+static MethodTable * g_pValueTypeMT;
+static WORD g_slotEquals;
+static WORD g_slotGetHashCode;
+
+static BOOL HasOverriddenEquals(MethodTable* mt)
+{
+    auto hasCheckedOverriddenEquals = [](MethodTable* mt) -> BOOL {
+        return mt->HasCheckedOverriddenEquals();
+    };
+    auto setHasCheckedOverriddenEquals = [](MethodTable* mt) -> void {
+        mt->SetHasCheckedOverriddenEquals();
+    };
+    auto hasOverriddenEquals = [](MethodTable* mt) -> BOOL {
+        return mt->HasOverriddenEquals();
+    };
+    auto setHasOverriddenEquals = [](MethodTable* mt) -> void {
```

Comment:
Using function pointers will add CFG (control flow guard) cost on each call.  If we use a single bit as Jan suggested, could we check for both `Equals` and `GetHashCode` overrides at the same time and avoid function pointers?